ftw/testbrowser/core.py

```diff
@@ -1018,6 +1018,31 @@ def _normalize_url(self, url_or_object, view=None):
 
         return url
 
+    @staticmethod
+    def _correct_webdav_xml(html):
+        if isinstance(html, (unicode, str)) and 'xmlns:d="DAV:"' in html:
+            html = html.strip().replace(
+                '<D:href>', '<d:href>').replace('</D:href>', '</d:href>')
+
+        elif isinstance(html, StringIO):
+            html.seek(0)
+            found = False
+            for line in html:
+                if 'xmlns:d="DAV:"' in line:
+                    found = True
+                    break
+            if found:
+                html.seek(0)
+                offset = html.tell()
+                for line in html:
+                    if 'D:href' in line:
+                        html.seek(offset)
+                        html.write(line.replace(
```

Comment:
I don't like that you are mutating the `StringIO` object while iterating over it.
When hitting this code, the html data will be parsed later by lxml, resulting in an lxml document which is kept completely in RAM. Since this will happen eventually we don't have to take care of not putting everything into RAM here.

Therefore I suggest to read the complete string into RAM, replace it, then recreate the StringIO. That should be better readable, simpler, less error prone but it may be consuming more RAM, but we don't care.

For example:
```python
elif isinstance(html, StringIO):
    position = html.tell()
    raw_html = html.getvalue()
    if 'xmlns:d="DAV:"' in raw_html:
        raw_html = raw_html.strip().replace('<D:href>', '<d:href>').replace('</D:href>', '</d:href>')
        html = String(raw_html)
        html.seek(position)
    return html
```

Reply:
Right. In that case why even bother to cast it back to StringIO? Parsing the html data is anyway the next thing that happens, so this might as well already be a string.